tiledb/api/c_api/group/group_api.cc

```diff
@@ -0,0 +1,797 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the group C API of TileDB.
+ **/
+
+#include "tiledb/api/c_api_support/c_api_support.h"
+#include "tiledb/sm/c_api/api_argument_validator.h"
+#include "tiledb/sm/c_api/tiledb_serialization.h"
+#include "tiledb/sm/c_api/tiledb_struct_def.h"
+#include "tiledb/sm/enums/serialization_type.h"
+#include "tiledb/sm/group/group_v1.h"
+#include "tiledb/sm/serialization/array.h"
+#include "tiledb/sm/serialization/group.h"
+
+#include "../config/config_api_internal.h"
+#include "../context/context_api_internal.h"
+
+#include "group_api_internal.h"
+
+namespace tiledb::api {
+
+tiledb::sm::URI ensure_group_uri_argument_is_valid(const char* group_uri, const bool& get_abs = true) {
+  if (group_uri == nullptr) {
+    throw CAPIStatusException("argument `group_uri` may not be nullptr");
+  }
+  auto uri = tiledb::sm::URI(group_uri, get_abs);
+  if (uri.is_invalid()) {
+    throw CAPIStatusException("argument `group_uri` is not a valid URI");
+  }
+  return uri;
+}
+
+capi_return_t tiledb_group_create(tiledb_ctx_handle_t* ctx, const char* group_uri) {
+  ensure_context_is_valid(ctx);
+  ensure_group_uri_argument_is_valid(group_uri);
+
+  throw_if_not_ok(ctx->storage_manager()->group_create(group_uri));
+
+  return TILEDB_OK;
+}
+
+capi_return_t tiledb_group_alloc(
+    tiledb_ctx_handle_t* ctx, const char* group_uri, tiledb_group_handle_t** group) {
+  ensure_context_is_valid(ctx);
+  ensure_output_pointer_is_valid(group);
+
+  auto uri = ensure_group_uri_argument_is_valid(group_uri);
+  *group = tiledb_group_handle_t::make_handle(
+    tdb_new(tiledb::sm::GroupV1, uri, ctx->storage_manager()));
+
+  return TILEDB_OK;
+}
+
+void tiledb_group_free(tiledb_group_handle_t** group) {
+  ensure_output_pointer_is_valid(group);
+  ensure_group_is_valid(*group);
+  tiledb_group_handle_t::break_handle(*group);
```

Comment:
Assuming I understand your meaning, this is fixed because I changed how the group handle allocates. I think you were basically just saying "This code is correct, but broken because of the other location". Other location is fixed, so I'm pretty sure this is resolved but will leave it open in case I've missed something subtle.